Sources/TuistLoader/Models+ManifestMappers/Headers+ManifestMapper.swift

```diff
@@ -12,58 +12,142 @@ extension TuistGraph.Headers {
-                    return false
-                }
-                return TuistGraph.Headers.extensions.contains(".\(fileExtension)") && !excluding.contains($0)
-            }
-        }
+        let resolvedUmbrellaPath = try manifest.umbrellaHeader.map { try generatorPaths.resolve(path: $0) }
+        let headersFromUmbrella = try manifest.publicHeadersFromUmbrella(resolvedPath: resolvedUmbrellaPath)
 
         var autoExlcudedPaths = Set<AbsolutePath>()
         let publicHeaders: [AbsolutePath]
         let privateHeaders: [AbsolutePath]
         let projectHeaders: [AbsolutePath]
 
-        func resolveHeaders(_ list: FileList?) throws -> [AbsolutePath] {
+        func unfold(_ list: FileList?, headersFromUmbrella: Set<String>? = nil) throws -> [AbsolutePath] {
             guard let list = list else { return [] }
-            return try list.globs.flatMap {
-                unfoldGlob(
-                    try generatorPaths.resolve(path: $0.glob),
-                    excluding: (try resolveExcluding($0.excluding)).union(autoExlcudedPaths)
-                )
+            var result = try list.unfold(
+                generatorPaths: generatorPaths,
+                basenameFilter: headersFromUmbrella,
+                additionalExcluding: autoExlcudedPaths
+            )
+            // be sure, that umbrella already here (if we use it)
+            if headersFromUmbrella != nil,
+               let resolvedUmbrellaPath = resolvedUmbrellaPath,
+               !result.contains(resolvedUmbrellaPath)
+            {
+                result.append(resolvedUmbrellaPath)
             }
+            return result
         }
 
         switch manifest.exclusionRule {
         case .projectExcludesPrivateAndPublic:
-            publicHeaders = try resolveHeaders(manifest.public)
+            publicHeaders = try unfold(manifest.public, headersFromUmbrella: headersFromUmbrella)
             autoExlcudedPaths.formUnion(publicHeaders)
-            privateHeaders = try resolveHeaders(manifest.private)
+            privateHeaders = try unfold(manifest.private)
             autoExlcudedPaths.formUnion(privateHeaders)
-            projectHeaders = try resolveHeaders(manifest.project)
+            projectHeaders = try unfold(manifest.project)
 
         case .publicExcludesPrivateAndProject:
-            projectHeaders = try resolveHeaders(manifest.project)
+            projectHeaders = try unfold(manifest.project)
             autoExlcudedPaths.formUnion(projectHeaders)
-            privateHeaders = try resolveHeaders(manifest.private)
+            privateHeaders = try unfold(manifest.private)
             autoExlcudedPaths.formUnion(privateHeaders)
-            publicHeaders = try resolveHeaders(manifest.public)
+            publicHeaders = try unfold(manifest.public, headersFromUmbrella: headersFromUmbrella)
         }
 
         return Headers(public: publicHeaders, private: privateHeaders, project: projectHeaders)
     }
 }
+
+extension FileList {
```

Comment:
I'd move this one to `FileList` because it's not scoped to `Headers+ManifestMapper.swift`

Reply:
fixed